File: openclaw-plugin/src/router.ts
/**
 * claw-turbo core regex routing engine (TypeScript port).
 * Matches user messages against compiled regex patterns and returns
 * the matched route with captured groups.
 */

export interface RouteConfig {
  name: string;
  description?: string;
  patterns: string[];
  command: string;
  response_template?: string;
}

export interface RoutesConfig {
  routes: RouteConfig[];
}

export interface CompiledRoute {
  name: string;
  description: string;
  patterns: RegExp[];
  rawPatterns: string[];
  command: string;
  responseTemplate: string;
}

export interface MatchResult {
  route: CompiledRoute;
  captures: Record<string, string>;
  rawMessage: string;
  matchTimeUs: number;
}

/**
 * Render a template string by replacing {{key}} with values.
 */
export function renderTemplate(
  template: string,
  variables: Record<string, string>
): string {
  return template.replace(/\{\{(\s*\w+\s*)\}\}/g, (match, key) => {
    const trimmed = key.trim();
    return variables[trimmed] ?? match;
  });
}

/**
 * Compile route configs into regex patterns.
 */
export function compileRoutes(configs: RouteConfig[]): CompiledRoute[] {
  return configs.map((cfg) => ({
    name: cfg.name,
    description: cfg.description ?? "",
    patterns: cfg.patterns.map((p) => new RegExp(p, "i")),
    rawPatterns: cfg.patterns,
    command: cfg.command,
    responseTemplate: cfg.response_template ?? "",
  }));
}

/**
 * Match a message against compiled routes.
 * Returns the first match or null.
 */
export function matchMessage(
  routes: CompiledRoute[],
  message: string
): MatchResult | null {
  const start = performance.now();

  for (const route of routes) {
    for (const pattern of route.patterns) {
      const m = pattern.exec(message);
      if (m) {
        const elapsed = (performance.now() - start) * 1000; // microseconds
        const captures: Record<string, string> = {};
        if (m.groups) {
          for (const [key, val] of Object.entries(m.groups)) {
            if (val !== undefined) captures[key] = val;
          }
        }
        return {
          route,
          captures,
          rawMessage: message,
          matchTimeUs: elapsed,
        };
      }
    }
  }

  return null;
}

File: claw_turbo/router.py
"""Core regex matching engine for routing user messages to skills."""

import logging
import re
import threading
import time
from dataclasses import dataclass
from pathlib import Path

from .config import ConfigWatcher, Route, load_routes

logger = logging.getLogger("claw-turbo")


@dataclass
class MatchResult:
    route: Route
    captures: dict[str, str]
    raw_message: str
    match_time_us: float  # microseconds

    @property
    def rendered_command(self) -> str:
        return _render_template(self.route.command, self._template_vars())

    @property
    def rendered_response(self) -> str:
        return _render_template(self.route.response_template, self._template_vars())

    def _template_vars(self) -> dict[str, str]:
        return {"raw_message": self.raw_message, **self.captures}


def _render_template(template: str, variables: dict[str, str]) -> str:
    """Replace {{key}} placeholders with values."""
    def replacer(m: re.Match) -> str:
        key = m.group(1).strip()
        return variables.get(key, m.group(0))
    return re.sub(r"\{\{(\s*\w+\s*)\}\}", replacer, template)


class Router:
    """Thread-safe regex router with hot-reload support."""

    def __init__(self, routes_path: Path | str | None = None, watch: bool = False):
        from .config import DEFAULT_ROUTES_PATH
        self._path = Path(routes_path) if routes_path else DEFAULT_ROUTES_PATH
        self._lock = threading.RLock()
        self._routes: list[Route] = load_routes(self._path)
        self._watcher: ConfigWatcher | None = None

        if watch:
            self._watcher = ConfigWatcher(self._path, self._reload)
            self._watcher.start()

    def _reload(self):
        new_routes = load_routes(self._path)
        with self._lock:
            self._routes = new_routes
        logger.info("Router reloaded with %d routes", len(new_routes))

    @property
    def routes(self) -> list[Route]:
        with self._lock:
            return list(self._routes)

    def match(self, message: str) -> MatchResult | None:
        """Try to match a message against all routes. Returns first match or None."""
        start = time.perf_counter()
        with self._lock:
            routes = list(self._routes)

        for route in routes:
            for pattern in route.patterns:
                m = pattern.search(message)
                if m:
                    elapsed_us = (time.perf_counter() - start) * 1_000_000
                    result = MatchResult(
                        route=route,
                        captures=m.groupdict(),
                        raw_message=message,
                        match_time_us=elapsed_us,
                    )
                    logger.info(
                        "MATCHED [%s] in %.1fus: %s",
                        route.name, elapsed_us, message[:80],
                    )
                    return result

        elapsed_us = (time.perf_counter() - start) * 1_000_000
        logger.debug("NO MATCH in %.1fus: %s", elapsed_us, message[:80])
        return None

    def stop(self):
        if self._watcher:
            self._watcher.stop()
